src/IROperator.h

```diff
@@ -1907,6 +1909,51 @@ inline Expr likely_if_innermost(Expr e) {
                                 {e}, Internal::Call::PureIntrinsic);
 }
 
+
+/** Cast an expression to the halide type corresponding to the C++
+ * type T clamping to the minimum and maximum values of the result
+ * type. */
+template <typename T>
+Expr saturating_cast(Expr e) {
+    return saturating_cast(type_of<T>(), e);
+}
+
+/** Cast an expression to a new type, clamping to the minimum and
+ * maximum values of the result type. */
+inline Expr saturating_cast(Type t, Expr e) {
+    // For float to float, guarantee infinities are always pinned to range.
```

Comment:
This case was added because the test failed as it has something like the following in the double to float case:
    if (input_double > max_for_float) {
        result = MAX_FLT;
    }
So for +Inf, that pins the result. (It so happens that (float)MAX_DBL is +Inf.) Since the obvious test code disagrees with passing through infinity, I chose to pin infinities to range. This seems like a good invariant: the result of saturating_cast is always in range for the result type. The other choice for double to float is to map anything greater than FLT_MAX to +Inf. (Obviously this mirrors onto the negatives as well.)

Orthogonality to all types is a general principle in Halide. It also makes a fair bit of sense if one is dealing with float16s. I think I'd rather have coordinates and such end up at the edge of space than at infinity as I have doubts most code will deal with infinity reasonably.